chain/chain/tests/finality.rs

```diff
@@ -0,0 +1,516 @@
+use near_chain::test_utils::setup;
+use near_chain::{ChainStore, ChainStoreAccess, ChainStoreUpdate};
+use near_chain::{FinalityGadget, FinalityGadgetQuorums};
+use near_crypto::{Signature, Signer};
+use near_primitives::block::{Approval, Block};
+use near_primitives::hash::CryptoHash;
+use near_primitives::types::{AccountId, BlockIndex};
+use near_store::test_utils::create_test_store;
+use rand::seq::SliceRandom;
+use rand::Rng;
+use std::collections::{HashMap, HashSet};
+
+fn compute_quorums_slow(
+    mut prev_hash: CryptoHash,
+    approvals: Vec<Approval>,
+    chain_store: &mut dyn ChainStoreAccess,
+    total_block_producers: usize,
+) -> FinalityGadgetQuorums {
+    let mut all_approvals = approvals;
+
+    let mut quorum_pre_vote = CryptoHash::default();
+    let mut quorum_pre_commit = CryptoHash::default();
+
+    let mut all_heights_and_hashes = vec![];
+
+    while prev_hash != CryptoHash::default() {
+        let block_header = chain_store.get_block_header(&prev_hash).unwrap();
+
+        all_heights_and_hashes.push((block_header.hash().clone(), block_header.inner.height));
```

Comment:
nit: hashes don't need to be cloned, they implement copy. If it's a reference, use `*hash()`